### ISSUE TITLE
Filehandle opened from ref to ref hangs on reading

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13213]
Migrated from [rt.perl.org#119529](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=119529) (status was 'resolved')

Searchable as RT119529$

**Comment by p5pRT** [comment]
### From @epa

#### Created by @epa



This causes perl to hang chewing CPU&#8203;:

&nbsp;    my $x = \\42;
&nbsp;    open my $fh\, "\<"\, \\$x;
&nbsp;    my $got = \<$fh>;

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.16.3:

Configured by Red Hat, Inc. at Fri May  3 12:10:03 UTC 2013.

Summary of my perl5 (revision 5 version 16 subversion 3) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-358.2.1.el6.x86_64, archname=x86_64-linux-thread-multi
    uname='linux buildvm-24.phx2.fedoraproject.org 2.6.32-358.2.1.el6.x86_64 #1 smp wed feb 20 12:17:37 est 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4  -m64 -mtune=generic -Dccdlflags=-Wl,--enable-new-dtags -Dlddlflags=-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4  -m64 -mtune=generic -Wl,-z,relro  -DDEBUGGING=-g -Dversion=5.16.3 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5 -Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl -Darchlib=/usr/lib64/perl5 -Dvendorarch=/usr/lib64/perl5/vendor_perl -Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Duseshrplib -Dusethreads -Duseithreads -Dusedtrace=/usr/bin/dtrace -Duselargefiles -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin -Dusesitecustomize'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2 20121109 (Red Hat 4.7.2-8)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.16'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Wl,-z,relro '

Locally applied patches:
    


@INC for perl 5.16.3:
    /home/eda/lib/perl5/
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib64/perl5
    /usr/share/perl5
    .


Environment for perl 5.16.3:
    HOME=/home/tradingsystems
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_GB.UTF-8
    LC_MESSAGES=en_GB.UTF-8
    LC_MONETARY=en_GB.UTF-8
    LC_NUMERIC=en_GB.UTF-8
    LC_TIME=en_GB.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/eda/bin:/home/eda/bin:/usr/local/bin:/usr/bin:/sbin:/usr/sbin:/sbin:/usr/sbin
    PERL5LIB=/home/eda/lib/perl5/
    PERL_BADLANG (unset)
    SHELL=/bin/bash


______________________________________________________________________
This email has been scanned by the Symantec Email Security.cloud service.
For more information please visit http://www.symanteccloud.com
______________________________________________________________________

```
</details>

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

On Fri Aug 30 10&#8203;:48&#8203;:16 2013\, eda@&#8203;waniasset\.com wrote&#8203;:
> hang chewing CPU&#8203;:

and RAM\, at least under 5\.19\.3

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @wolfsage

On Fri\, Aug 30\, 2013 at 1&#8203;:48 PM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

>
> This causes perl to hang chewing CPU&#8203;:
>
>     my $x = \\42;
>     open my $fh\, "\<"\, \\$x;
>     my $got = \<$fh>;
>
>
This seems to be in sv\.c\, in Perl\_sv\_gets\, in the loop starting here\, at
least for me&#8203;:

8097  for \(;;\) \{
8098       screamer&#8203;:
8099         if \(cnt > 0\) \{
8100             if \(rslen\) \{
8101                 while \(cnt > 0\) \{                    /\* this     |
eat \*/
8102                     cnt\-\-;
8103                     if \(\(\*bp\+\+ = \*ptr\+\+\) == rslast\)  /\* really   |
dust \*/
8104                         goto thats\_all\_folks;        /\* screams  |
sed :\-\) \*/
8105                 \}
8106             \}
8107             else \{
8108                 Copy\(ptr\, bp\, cnt\, char\);            /\* this     |
eat \*/
8109                 bp \+= cnt;                           /\* screams  |
dust \*/
8110                 ptr \+= cnt;                          /\* louder   |
sed :\-\) \*/
8111                 cnt = 0;
8112                 assert \(\!shortbuffered\);
8113                 goto cannot\_be\_shortbuffered;
8114             \}
8115         \}
8116

I haven't chased it any further\. I didn't see it on perl\-5\.10\.1\, but I do
see it on perl\-5\.14\.2\.

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 30 12&#8203;:46&#8203;:31 2013\, alh wrote&#8203;:
> On Fri\, Aug 30\, 2013 at 1&#8203;:48 PM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org>
wrote&#8203;:
> 
> >
> > This causes perl to hang chewing CPU&#8203;:
> >
> >     my $x = \\42;
> >     open my $fh\, "\<"\, \\$x;
> >     my $got = \<$fh>;
> >
> >
> This seems to be in sv\.c\, in Perl\_sv\_gets\, in the loop starting here\, at
> least for me&#8203;:
> 
> 8097  for \(;;\) \{
> 8098       screamer&#8203;:
> 8099         if \(cnt > 0\) \{
> 8100             if \(rslen\) \{
> 8101                 while \(cnt > 0\) \{                    /\* this     |
> eat \*/
> 8102                     cnt\-\-;
> 8103                     if \(\(\*bp\+\+ = \*ptr\+\+\) == rslast\)  /\* really   |
> dust \*/
> 8104                         goto thats\_all\_folks;        /\* screams  |
> sed :\-\) \*/
> 8105                 \}
> 8106             \}
> 8107             else \{
> 8108                 Copy\(ptr\, bp\, cnt\, char\);            /\* this     |
> eat \*/
> 8109                 bp \+= cnt;                           /\* screams  |
> dust \*/
> 8110                 ptr \+= cnt;                          /\* louder   |
> sed :\-\) \*/
> 8111                 cnt = 0;
> 8112                 assert \(\!shortbuffered\);
> 8113                 goto cannot\_be\_shortbuffered;
> 8114             \}
> 8115         \}
> 8116
> 
> I haven't chased it any further\. I didn't see it on perl\-5\.10\.1\, but I do
> see it on perl\-5\.14\.2\.
> 
> \-\- Matthew Horsfall \(alh\)

$ \.\./perl\.git/Porting/bisect\.pl \-\-start=v5\.10\.0 \-\-end=v5\.12\.4 \-e 'alarm
1; my $x = \\42; open my $fh\, "\<"\, \\$x; my $got = \<$fh>;'
\.\.\.
30c16bb48be7651b70d0995a08678033f9a5e942 is the first bad commit
commit 30c16bb48be7651b70d0995a08678033f9a5e942
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Wed May 5 22&#8203;:39&#8203;:24 2010 \+0100

&nbsp;    RT 43789&#8203;: "in memory" files don't call STORE
&nbsp;    
&nbsp;    The code in PerlIO\-scalar that implements the open $fh\, '>' \\$buffer
&nbsp;    feature did not\, apart from accidentally\, support get/set magic and thus
&nbsp;    tied buffers\. This patch remedies that&#8203;: mostly by just blindly
sprinkling
&nbsp;    SvGETMAGIC/SvSETMAGIC about\, rather than doing any deep analysis and
&nbsp;    understanding of the code\. One main change I did was to add a
&nbsp;    PerlIOScalar\_read\(\) function\, rather than rely on the default behaviour
&nbsp;    \(which implements it in terms of PerlIOScalar\_get\_ptr\(\) etc\)\, since that
&nbsp;    approach had a tendency to call FETCH multiple times

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @epa

In general I suggest that opening a filehandle giving a reference to a reference doesn't make any sense\, and surely didn't do anything useful even before the change identified\.  It would be better to make it throw an exception\.

Similarly I suggest that open $fh\, '\<'\, \\undef should also throw\.

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Sep 23\, 2013 at 06&#8203;:59&#8203;:09AM \+0000\, Ed Avis wrote&#8203;:
> In general I suggest that opening a filehandle giving a reference to a
> reference doesn't make any sense\, and surely didn't do anything useful
> even before the change identified\.  It would be better to make it throw
> an exception\.

The other approach is to treat it similarly to how other non\-string vars
are treated when used as strings\, e\.g\. $x = \\1; $x \.= 2\, which sets
$x to "SCALAR\(0x11fd5b0\)2"\.

> Similarly I suggest that open $fh\, '\<'\, \\undef should also throw\.

The test suite for PerlIO&#8203;::scalar already explicitly expects undef
to behave like an empty string \*without\* generating any warnings\.

I've pushed the following\, which fixes the original bug report by going
down the stringifying route\.


commit 552908b17471877454d99aa32f7ac5c29ea2efbb
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Thu Oct 17 15&#8203;:35&#8203;:14 2013 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Thu Oct 17 15&#8203;:46&#8203;:05 2013 \+0100

&nbsp;    PerlIO&#8203;::scalar&#8203;: stringify refs
&nbsp;    
&nbsp;    If $s in
&nbsp;        open my $fh\, "\<"\,  \\$s
&nbsp;    or similar is a ref\, then
&nbsp;    stringify that ref&#8203;: i\.e\. convert it from a ref into the string
&nbsp;    "SCALAR\(0x\.\.\.\.\)" or whatever\.
&nbsp;    
&nbsp;    This fixes
&nbsp;    
&nbsp;        RT \#119529 Filehandle opened from ref to ref hangs on reading
&nbsp;    
&nbsp;    which in this case was looping forever\, since it kept thinking that
&nbsp;    the var was a string \("SCALAR\.\."\)\, but whose length was 0\.
&nbsp;    
&nbsp;    I haven't gone for a complete "always force var into a string" approach\,
&nbsp;    since PerlIO&#8203;::scalar has quite a tolerance for "bad" vars; e\.g\.
&nbsp;    it won't warn if $var is undef or a read\-only constant number etc;
&nbsp;    and it already normalises under some circumstances and not others\.
&nbsp;    So I've just increased the cases somewhat where it normalises\.
&nbsp;    
&nbsp;    Also\, I didn't look to closely at the code that was looping \(or to put it
&nbsp;    another way\, I looked at it but didn't understand it\)\, so it could
&nbsp;    conceivably still behave badly on some other strange type of variable that
&nbsp;    manages to avoid getting normalised\.


\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." \-\- Lady Croom\, "Arcadia"

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @epa

Thanks for fixing this\.  If silently treating undef as the empty string really is the desired behaviour\,
it would be worth documenting that\.  \(IMHO a warning would be more useful\, but I guess
whoever wrote the PerlIO&#8203;::scalar test suite already weighed up the pros and cons of that behaviour\.\) 

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Oct 17 07&#8203;:58&#8203;:17 2013\, davem wrote&#8203;:
> I've pushed the following\, which fixes the original bug report by
> going
> down the stringifying route\.
> 
> 
> commit 552908b17471877454d99aa32f7ac5c29ea2efbb
> Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
> AuthorDate&#8203;: Thu Oct 17 15&#8203;:35&#8203;:14 2013 \+0100
> Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
> CommitDate&#8203;: Thu Oct 17 15&#8203;:46&#8203;:05 2013 \+0100

\.\.\.which unfortunately breaks when reading a read\-only ref in "\<" mode\.  I pushed a better fix as 5a2bc23b\.

I actually wrote something similar to that a month ago\, but then sat on it as real life got in the way\.

The amazing thing was that both Dave Mitchell and I used \\42 it our tests\.  Great minds think alike\!

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Oct 25\, 2013 at 02&#8203;:23&#8203;:10PM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Thu Oct 17 07&#8203;:58&#8203;:17 2013\, davem wrote&#8203;:
> > I've pushed the following\, which fixes the original bug report by
> > going
> > down the stringifying route\.
> > 
> > 
> > commit 552908b17471877454d99aa32f7ac5c29ea2efbb
> > Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
> > AuthorDate&#8203;: Thu Oct 17 15&#8203;:35&#8203;:14 2013 \+0100
> > Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
> > CommitDate&#8203;: Thu Oct 17 15&#8203;:46&#8203;:05 2013 \+0100
> 
> \.\.\.which unfortunately breaks when reading a read\-only ref in "\<" mode\.  I pushed a better fix as 5a2bc23b\.

Ok thanks\.

> The amazing thing was that both Dave Mitchell and I used \\42 it our
> tests\.  Great minds think alike\!

Er\, actually the \\42 was in the sample code in the original ticket\!

\-\- 
That he said that that that that is is is debatable\, is debatable\.
